src/shared/components/download-button/DownloadButton.tsx

```diff
@@ -15,34 +15,106 @@
  */
 
 import React, {
+  useState,
   forwardRef,
-  type HTMLAttributes,
+  type ComponentProps,
   type ForwardedRef
 } from 'react';
 import classNames from 'classnames';
 
+import { CircleLoader } from 'src/shared/components/loader';
+
 import DownloadIcon from 'static/icons/icon_download.svg';
+import Checkmark from 'static/icons/icon_tick.svg';
+import Cross from 'static/icons/icon_cross.svg';
+
+import { LoadingState } from 'src/shared/types/loading-state';
 
 import styles from './DownloadButton.scss';
 
-type Props = Omit<HTMLAttributes<HTMLButtonElement>, 'children'> & {
-  disabled?: boolean;
+type Props = Omit<ComponentProps<'button'>, 'children' | 'onClick'> & {
+  onClick: () => unknown;
 };
 
 const DownloadButton = (props: Props, ref: ForwardedRef<HTMLButtonElement>) => {
-  const { className, ...otherProps } = props;
+  const [status, setStatus] = useState(LoadingState.NOT_REQUESTED);
+  const { onClick, ...otherProps } = props;
+
+  const clickHandler = async () => {
+    setStatus(LoadingState.LOADING);
+    try {
+      await onClick();
+      setStatus(LoadingState.SUCCESS);
+    } catch {
+      setStatus(LoadingState.ERROR);
+    } finally {
+      setTimeout(() => {
+        setStatus(LoadingState.NOT_REQUESTED);
+      }, 1000);
+    }
+  };
+
+  return (
+    <ControlledDownloadButtonWithForwardedRef
+      {...otherProps}
+      ref={ref}
+      status={status}
+      onClick={clickHandler}
+    />
+  );
+};
+
+type ControlledProps = Props & {
+  status: LoadingState;
+};
+
+const ControlledDownloadButton = (
+  props: ControlledProps,
+  ref: ForwardedRef<HTMLButtonElement>
+) => {
+  const { className, status, ...otherProps } = props;
 
   const elementClasses = classNames(
     styles.downloadButton,
-    { [styles.downloadButtonDisabled]: props.disabled },
+    {
+      [styles.downloadButtonDisabled]: props.disabled,
+      [styles.downloadButtonLoading]: status === LoadingState.LOADING,
+      [styles.downloadButtonSuccess]: status === LoadingState.SUCCESS,
+      [styles.downloadButtonError]: status === LoadingState.ERROR
+    },
     className
   );
 
+  const isDisabled = status !== LoadingState.NOT_REQUESTED || props.disabled;
+
   return (
-    <button {...otherProps} ref={ref} className={elementClasses}>
-      <DownloadIcon />
+    <button
+      {...otherProps}
+      ref={ref}
+      className={elementClasses}
+      disabled={isDisabled}
+    >
+      {getButtonContent(status)}
     </button>
   );
 };
 
+const getButtonContent = (status: LoadingState) => {
+  switch (status) {
+    case LoadingState.LOADING:
+      return <CircleLoader size="small" />;
+    case LoadingState.SUCCESS:
+      return <Checkmark />;
+    case LoadingState.ERROR:
+      return <Cross />;
+    default:
+      return <DownloadIcon />;
+  }
+};
+
+const ControlledDownloadButtonWithForwardedRef = forwardRef(
+  ControlledDownloadButton
+);
+
+export { ControlledDownloadButtonWithForwardedRef as ControlledDownloadButton };
```

Comment:
Why do we need to support/use `ref` in this component?

Reply:
`DownloadButton` is a shared component that can be used in various places; so I am trying to treat it as if it were just a regular HTML button. Which means that, since I can access a regular `button` element in the DOM by passing a `ref` to a React component representing it, I want to also be able to access the `button` element via a ref passed to the `DownloadButton` component. It is a useful thing to be able to do when you need it; and it's easy to set up.